modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java

```diff
@@ -2749,107 +2822,105 @@ private double getPrefLength() {
-        // of the corner cases and provides a simpler algorithm without adding
-        // any overhead to performance.
-        double start = computeOffsetForCell(cellIndex);
-        double end = cellSize + computeOffsetForCell(cellIndex + 1);
-
-        // We need to discover the distance that the fictional "position line"
-        // would travel within this cell, from its current position to the end.
-        double remaining = end - start;
-
-        // Keep track of the number of pixels left to travel
-        double n = forward ?
-              numPixels + pixelOffset - (getViewportLength() * getPosition()) - start
-            : -numPixels + end - (pixelOffset - (getViewportLength() * getPosition()));
-
-        // "p" represents the most recent value for position. This is always
-        // based on the edge between two cells, except at the very end of the
-        // algorithm where it is added to the computed "p" offset for the final
-        // value of Position.
-        double p = cellPercent * cellIndex;
-
-        // Loop over the cells one at a time until either we reach the end of
-        // the cells, or we find that the "n" will fall within the cell we're on
-        while (n > remaining && ((forward && cellIndex < cellCount - 1) || (! forward && cellIndex > 0))) {
-            if (forward) cellIndex++; else cellIndex--;
-            n -= remaining;
-            cellSize = getCellLength(cellIndex);
-            start = computeOffsetForCell(cellIndex);
-            end = cellSize + computeOffsetForCell(cellIndex + 1);
-            remaining = end - start;
-            p = cellPercent * cellIndex;
-        }
-
-        // if remaining is < n, then we must have hit an end, so as a
-        // fast path, we can just set position to 1.0 or 0.0 and return
-        // because we know we hit the end
-        if (n > remaining) {
-            setPosition(forward ? 1.0f : 0.0f);
-        } else if (forward) {
-            double rate = cellPercent / Math.abs(end - start);
-            setPosition(p + (rate * n));
-        } else {
-            double rate = cellPercent / Math.abs(end - start);
-            setPosition((p + cellPercent) - (rate * n));
+    private double adjustByPixelAmount(double numPixels) {
+        if (numPixels == 0) return 0;
+        // When we're at the top already, we can't move back further, unless we
+        // want to allow for gravity-alike effects.
+        if ((absoluteOffset <= 0) && (numPixels < 0)) return 0;
+
+        // start with applying the requested modification
+        double origAbsoluteOffset = this.absoluteOffset;
+        this.absoluteOffset = Math.max(0.d, this.absoluteOffset + numPixels);
+        double newPosition = Math.min(1.0d, absoluteOffset/(estimatedSize - viewportLength));
+        // estimatedSize changes may result in opposite effect on position
+        // in that case, modify current position 1% in the requested direction
```

Comment:
Have you done testing to show that a 1% adjustment is a good heuristic?

Reply:
This is just a visual aid, so it's hard to test. I tried this with both small and large lists. We really don't want to jump in the wrong direction, but on the other hand, moving too much in the "correct" dimension would increase the discrepancy between where we really are in the list versus where we expected to be, given the information we had before this modification. If we increase this to 10%, the mismatch will be even bigger, and might be visually noticed. 

The 1% is a compromis between the UI going crazy (backward) and the mathematical correctness (which would go backward). I initially had this to 0% but then there would be no visual change, while we want to give the user some visual confirmation about the direction she is moving. 
With both small/large lists, this 1% occurs "natural" to me. The lower, the better, but if we change less than 1 pixel, there won't be a visual effect.